src/libsystemd/sd-device/device-enumerator.c

```diff
@@ -249,26 +248,22 @@ static int device_compare(sd_device * const *_a, sd_device * const *_b) {
 
                         prefix_len = sound_a - devpath_a;
 
-                        if (strncmp(devpath_a, devpath_b, prefix_len) == 0) {
+                        if (strneq(devpath_a, devpath_b, prefix_len)) {
                                 const char *sound_b;
 
                                 sound_b = devpath_b + prefix_len;
 
-                                if (startswith(sound_a, "/controlC") &&
-                                    !startswith(sound_b, "/contolC"))
-                                        return 1;
-
-                                if (!startswith(sound_a, "/controlC") &&
-                                    startswith(sound_b, "/controlC"))
-                                        return -1;
+                                r = CMP(!!startswith(sound_a, "/controlC"),
+                                        !!startswith(sound_b, "/controlC"));
+                                if (r != 0)
+                                        return r;
                         }
                 }
         }
 
         /* md and dm devices are enumerated after all other devices */
-        delay_a = strstr(devpath_a, "/block/md") || strstr(devpath_a, "/block/dm-");
-        delay_b = strstr(devpath_b, "/block/md") || strstr(devpath_b, "/block/dm-");
-        r = CMP(delay_a, delay_b);
+        r = CMP(strstr(devpath_a, "/block/md") || strstr(devpath_a, "/block/dm-"),
+                strstr(devpath_b, "/block/md") || strstr(devpath_b, "/block/dm-"));
```

Comment:
if you touch this, maybe we should make this a tiny bit nicer with a helper:

```c
static bool syspath_is_late_block(const char *p) {
        return strstr(p, "/block/md") || strstr(p, "/block/dm-");
}
```

so that you then just call:

```c
r = CMP(syspath_is_late_block(devpath_a), syspath_is_late_block(devpath_b));
```

which is nicer on the eye and less duplication